src/TurbulenceConvection/EDMF_functions.jl

```diff
@@ -0,0 +1,1352 @@
+        tends_ρaq_tot = tendencies_up[i].ρaq_tot
+
+        @. tends_ρarea =
+            -∇c(wvec(LBF(Ic(w_up) * ρarea))) +
+            (ρarea * Ic(w_up) * entr_turb_dyn) -
+            (ρarea * Ic(w_up) * detr_turb_dyn)
+
+        @. tends_ρaθ_liq_ice =
+            -∇c(wvec(LBF(Ic(w_up) * ρaθ_liq_ice))) +
+            (ρarea * Ic(w_up) * entr_turb_dyn * θ_liq_ice_en) -
+            (ρaθ_liq_ice * Ic(w_up) * detr_turb_dyn) +
+            (ρ_c * θ_liq_ice_tendency_precip_formation)
+
+        @. tends_ρaq_tot =
+            -∇c(wvec(LBF(Ic(w_up) * ρaq_tot))) +
+            (ρarea * Ic(w_up) * entr_turb_dyn * q_tot_en) -
+            (ρaq_tot * Ic(w_up) * detr_turb_dyn) +
+            (ρ_c * qt_tendency_precip_formation)
+
+        if edmf.moisture_model isa NonEquilibriumMoisture
+
+            q_liq_up = aux_up_i.q_liq
+            q_ice_up = aux_up_i.q_ice
+            q_liq_en = aux_en.q_liq
+            q_ice_en = aux_en.q_ice
+
+            ql_tendency_noneq = aux_up_i.ql_tendency_noneq
+            qi_tendency_noneq = aux_up_i.qi_tendency_noneq
+            ql_tendency_precip_formation = aux_up_i.ql_tendency_precip_formation
+            qi_tendency_precip_formation = aux_up_i.qi_tendency_precip_formation
+
+            ρaq_liq = prog_up[i].ρaq_liq
+            ρaq_ice = prog_up[i].ρaq_ice
+
+            tends_ρaq_liq = tendencies_up[i].ρaq_liq
+            tends_ρaq_ice = tendencies_up[i].ρaq_ice
+
+            @. tends_ρaq_liq =
+                -∇c(wvec(LBF(Ic(w_up) * ρaq_liq))) +
+                (ρarea * Ic(w_up) * entr_turb_dyn * q_liq_en) -
+                (ρaq_liq * Ic(w_up) * detr_turb_dyn) +
+                (ρ_c * (ql_tendency_precip_formation + ql_tendency_noneq))
+
+            @. tends_ρaq_ice =
+                -∇c(wvec(LBF(Ic(w_up) * ρaq_ice))) +
+                (ρarea * Ic(w_up) * entr_turb_dyn * q_ice_en) -
+                (ρaq_ice * Ic(w_up) * detr_turb_dyn) +
+                (ρ_c * (qi_tendency_precip_formation + qi_tendency_noneq))
+
+            tends_ρaq_liq[kc_surf] = 0
+            tends_ρaq_ice[kc_surf] = 0
+        end
+
+        # prognostic entr/detr
+        if edmf.entr_closure isa PrognosticNoisyRelaxationProcess
+            c_gen_stoch = edmf.entr_closure.c_gen_stoch
+            mean_entr = aux_up[i].ε_nondim
+            mean_detr = aux_up[i].δ_nondim
+            ε_λ = c_gen_stoch[3]
+            δ_λ = c_gen_stoch[4]
+            tends_ε_nondim = tendencies_up[i].ε_nondim
+            tends_δ_nondim = tendencies_up[i].δ_nondim
+            ε_nondim = prog_up[i].ε_nondim
+            δ_nondim = prog_up[i].δ_nondim
+            @. tends_ε_nondim = ε_λ * (mean_entr - ε_nondim)
+            @. tends_δ_nondim = δ_λ * (mean_detr - δ_nondim)
+        end
+
+        tends_ρarea[kc_surf] = 0
+        tends_ρaθ_liq_ice[kc_surf] = 0
+        tends_ρaq_tot[kc_surf] = 0
+    end
+
+    # Solve for updraft velocity
+
+    # We know that, since W = 0 at z = 0, BCs for entr, detr,
```

Comment:
Here too: physical vertical velocity and covariant vertical velocities are nonzero at surfaces with topography. Revisit.